File: android/src/main/kotlin/dev/brewkits/native_workmanager/engine/FlutterEngineManager.kt
package dev.brewkits.native_workmanager.engine

import android.content.Context
import android.util.Log
import io.flutter.FlutterInjector
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import dev.brewkits.native_workmanager.workers.utils.ProgressReporter
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.FlutterCallbackInformation
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/**
 * Manages FlutterEngine lifecycle for Dart worker execution.
 *
 * This manager implements a singleton pattern with:
 * - Engine caching (reuse between tasks for performance)
 * - Auto-disposal after inactivity
 * - Thread-safe initialization
 * - Timeout handling
 *
 * Performance characteristics:
 * - First task (cold start): 500-1000ms (engine initialization)
 * - Subsequent tasks (engine cached): 100-200ms
 * - RAM: ~30-50MB while engine is alive
 * - Auto-dispose: 5 minutes after last task
 *
 * @see DartCallbackWorker
 */
object FlutterEngineManager {

    private const val TAG = "FlutterEngineManager"
    private const val CHANNEL_NAME = "dev.brewkits/dart_worker_channel"
    private const val ENGINE_IDLE_TIMEOUT_MS = 5 * 60 * 1000L // 5 minutes

    private var engine: FlutterEngine? = null
    private var methodChannel: MethodChannel? = null
    private var lastUsedTimestamp = 0L
    private val initializationMutex = Mutex()
    private val isInitialized = AtomicBoolean(false)

    // Callback handle stored during plugin initialization
    private var callbackHandle: Long? = null

    // Coroutine scope for auto-disposal management
    private val disposalScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var disposalJob: Job? = null

    /**
     * Set the Dart callback handle.
     *
     * This must be called during plugin initialization with the handle
     * of the top-level callback dispatcher function.
     */
    fun setCallbackHandle(handle: Long) {
        callbackHandle = handle
        Log.d(TAG, "Callback handle set: $handle")
    }

    // Reference counter (AtomicInteger) to prevent disposal while tasks are running.
    // Multiple coroutines on different threads can call executeDartCallback() concurrently;
    // plain var Int would race on ++ / -- (non-atomic read-modify-write).
    private val activeTaskCount = AtomicInteger(0)

    /**
     * Execute a Dart callback with input data.
     */
    suspend fun executeDartCallback(
        context: Context,
        callbackHandle: Long,
        input: String?,
        timeoutMs: Long = 5 * 60 * 1000L,
        disposeImmediately: Boolean = false
    ): Boolean = withContext(Dispatchers.Main) {
        try {
            Log.d(TAG, "Executing Dart callback with handle: $callbackHandle")

            ensureEngineInitialized(context)
            activeTaskCount.incrementAndGet()

            val channel = methodChannel
            if (channel == null) {
                activeTaskCount.decrementAndGet()
                return@withContext false
            }

            val resultDeferred = CompletableDeferred<Boolean>()
            val args = mapOf("callbackHandle" to callbackHandle, "input" to input)

            channel.invokeMethod("executeCallback", args, object : MethodChannel.Result {
                override fun success(result: Any?) {
                    resultDeferred.complete((result as? Boolean) ?: false)
                }
                override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
                    resultDeferred.complete(false)
                }
                override fun notImplemented() {
                    resultDeferred.complete(false)
                }
            })

            val result = try {
                withTimeout(timeoutMs) { resultDeferred.await() }
            } finally {
                activeTaskCount.decrementAndGet()
            }

            if (disposeImmediately && activeTaskCount.get() <= 0) {
                dispose()
            } else {
                lastUsedTimestamp = System.currentTimeMillis()
                scheduleDisposalCheck()
            }

            result
        } catch (e: Exception) {
            Log.e(TAG, "Error executing Dart callback", e)
            // Error occurred — likely a timeout or crash. Dispose immediately if no other tasks
            // are running to ensure a clean slate and free RAM from a potentially hung Isolate.
            if (activeTaskCount.get() <= 0) {
                try { dispose() } catch (_: Exception) {}
            }
            false
        }
    }

    /**
     * Ensure Flutter engine is initialized.
     *
     * This uses double-checked locking for thread safety and performance.
     */
    private suspend fun ensureEngineInitialized(context: Context) {
        // Fast path: already initialized
        if (isInitialized.get()) {
            return
        }

        // Slow path: need to initialize
        initializationMutex.withLock {
            // Double-check after acquiring lock
            if (isInitialized.get()) {
                return
            }

            Log.d(TAG, "Initializing Flutter engine...")

            val handle = callbackHandle ?: run {
                throw IllegalStateException(
                    "Callback handle not set. " +
                    "Call FlutterEngineManager.setCallbackHandle() during plugin initialization."
                )
            }

            // Get callback information
            val callbackInfo = FlutterCallbackInformation.lookupCallbackInformation(handle)
            if (callbackInfo == null) {
                throw IllegalStateException("Failed to lookup callback information for handle: $handle")
            }

            Log.d(TAG, "Callback info: ${callbackInfo.callbackName}")

            // Create engine
            val flutterEngine = FlutterEngine(context.applicationContext)

            // Destroy the engine on any init failure so we don't leak
            // ~30-50MB of RAM per failed attempt (invalid callback handle, 10s timeout, etc.).
            try {
                // Get the correct app bundle path for the Flutter assets.
                // context.assets.toString() returns the AssetManager object string, which is wrong.
                // FlutterLoader provides the correct path for both debug (JIT) and release (AOT) modes.
                val bundlePath = FlutterInjector.instance().flutterLoader().findAppBundlePath()

                val dartCallback = DartExecutor.DartCallback(
                    context.assets,
                    bundlePath,
                    callbackInfo
                )

                // Create MethodChannel and register 'dartReady' handler BEFORE starting the Dart isolate.
                // This eliminates the race condition where Dart sends 'dartReady' before the Kotlin
                // handler is registered, which would cause the message to be dropped and the
                // waitForDartReady() 10-second timeout to fire.
                val channel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL_NAME)
                val readyDeferred = CompletableDeferred<Unit>()
                channel.setMethodCallHandler { call, result ->
                    when (call.method) {
                        "dartReady" -> {
                            Log.d(TAG, "Dart side ready")
                            readyDeferred.complete(Unit)
                            result.success(null)
                        }
                        // Progress reports emitted from inside a DartWorker callback.
                        // The Dart side calls MethodChannel('dev.brewkits/dart_worker_channel')
                        // .invokeMethod('reportProgress', {...}) which arrives here and is
                        // forwarded to the shared ProgressReporter → Flutter EventChannel.
                        "reportProgress" -> {
                            val taskId  = call.argument<String>("taskId") ?: ""
                            val progress = call.argument<Int>("progress") ?: 0
                            val message  = call.argument<String>("message")
                            ProgressReporter.reportProgressNonBlocking(taskId, progress, message)
                            result.success(null)
                        }
                        else -> result.notImplemented()
                    }
                }

                // Now start the Dart isolate — handler is already registered, no race possible.
                flutterEngine.dartExecutor.executeDartCallback(dartCallback)

                // Wait for Dart to signal it's ready (up to 10 seconds)
                withTimeout(10_000L) {
                    readyDeferred.await()
                }
                Log.d(TAG, "Dart ready signal received")

                // Store references only after successful initialization
                engine = flutterEngine
                methodChannel = channel
                lastUsedTimestamp = System.currentTimeMillis()
                isInitialized.set(true)

                Log.d(TAG, "Flutter engine initialized successfully")
            } catch (e: Exception) {
                // DART-007: Ensure engine is destroyed on any failure path to prevent memory leak.
                withContext(Dispatchers.Main) {
                    try { flutterEngine.destroy() } catch (_: Exception) {}
                }
                throw e
            }
        }
    }

    /**
     * Schedule engine disposal check.
     *
     * If engine hasn't been used for ENGINE_IDLE_TIMEOUT_MS,
     * it will be disposed to free memory.
     */
    private fun scheduleDisposalCheck() {
        // Auto-disposal with proper CoroutineScope
        // Uses SupervisorJob to prevent leaks and properly managed lifecycle

        // Cancel any existing disposal job
        disposalJob?.cancel()

        // Schedule new disposal check
        disposalJob = disposalScope.launch {
            delay(ENGINE_IDLE_TIMEOUT_MS)

            // Check if engine is still idle
            val idleTime = System.currentTimeMillis() - lastUsedTimestamp
            if (idleTime >= ENGINE_IDLE_TIMEOUT_MS) {
                Log.d(TAG, "Auto-disposing engine after ${idleTime}ms idle")
                disposalScope.launch { dispose() }
            }
        }
    }

    /**
     * Dispose the Flutter engine.
     *
     * Acquires [initializationMutex] to prevent a concurrent
     * [ensureEngineInitialized] call from storing a reference to a destroyed
     * engine, or vice-versa (H-2 fix).
     *
     * This frees ~30-50MB of RAM but future tasks will be slower
     * (need to cold-start engine again).
     */
    suspend fun dispose() {
        initializationMutex.withLock {
            Log.d(TAG, "Disposing Flutter engine")

            methodChannel?.setMethodCallHandler(null)
            methodChannel = null

            try {
                withContext(Dispatchers.Main) {
                    engine?.destroy()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error destroying engine (expected if already detached)", e)
            }
            engine = null

            isInitialized.set(false)

            // Cancel disposal job to prevent unnecessary work
            disposalJob?.cancel()
            disposalJob = null

            Log.d(TAG, "Flutter engine disposed")
        }
    }

    /**
     * Check if engine is currently initialized.
     */
    fun isEngineAlive(): Boolean = isInitialized.get()

    /**
     * Get time since engine was last used (for monitoring).
     */
    fun getIdleTimeMs(): Long {
        if (!isInitialized.get()) return -1
        return System.currentTimeMillis() - lastUsedTimestamp
    }
}

File: ios/Classes/engine/FlutterEngineManager.swift
import Flutter
import Foundation

/// Singleton Flutter Engine manager for iOS Dart workers.
///
/// Manages the lifecycle of a Flutter Engine used to execute Dart callbacks
/// in background tasks. Implements engine caching for performance:
/// - First execution (cold start): 500-1000ms
/// - Subsequent executions (cached): 100-200ms
///
/// **Thread Safety:** All public methods are thread-safe and can be called
/// from any queue/thread.
///
/// **Memory:** Engine consumes 30-50MB RAM when active. Designed for
/// auto-disposal after idle timeout (not yet implemented).
class FlutterEngineManager {

    // MARK: - Singleton

    static let shared = FlutterEngineManager()

    private init() {}

    // MARK: - Properties

    private var engine: FlutterEngine?
    private var methodChannel: FlutterMethodChannel?
    private var callbackHandle: Int64?

    private let queue = DispatchQueue(label: "dev.brewkits.flutter_engine_manager")
    private var isInitialized = false
    private var isInitializing = false
    private var initializationContinuations: [CheckedContinuation<Void, Error>] = []

    // Timeout for engine initialization and callback execution
    private static let initTimeoutSeconds: TimeInterval = 30
    private static let defaultCallbackTimeoutSeconds: TimeInterval = 300 // 5 minutes

    // ✅ IMPLEMENTED: Auto-disposal after idle timeout
    private var lastUsedTimestamp: Date?
    private static let idleTimeoutSeconds: TimeInterval = 300 // 5 minutes
    private var disposalWorkItem: DispatchWorkItem?

    // MARK: - Public API

    /// Set the callback handle from Dart side.
    ///
    /// Must be called during plugin initialization before executing any Dart callbacks.
    ///
    /// - Parameter handle: Callback handle from PluginUtilities.getCallbackHandle()
    func setCallbackHandle(_ handle: Int64) {
        queue.sync {
            self.callbackHandle = handle
            print("FlutterEngineManager: Callback handle registered: \(handle)")
        }
    }

    /// Check if Flutter Engine is currently alive.
    var isEngineAlive: Bool {
        queue.sync {
            return engine != nil && isInitialized
        }
    }

    /// Execute a Dart callback in the background isolate.
    ///
    /// This method ensures the engine is initialized, then invokes the specified
    /// Dart callback with the given input. Supports aggressive disposal to free memory immediately.
    ///
    /// - Parameters:
    ///   - callbackHandle: The callback handle from PluginUtilities.getCallbackHandle()
    ///   - input: JSON string input for the callback
    ///   - timeoutSeconds: Maximum time to wait for callback completion
    ///   - disposeImmediately: If true, engine is killed immediately after callback completes (saves ~50MB RAM)
    /// - Returns: true if callback succeeded, false otherwise
    /// - Throws: Errors related to engine initialization or timeout
    func executeDartCallback(
        callbackHandle: Int64,
        input: String?,
        timeoutSeconds: TimeInterval = defaultCallbackTimeoutSeconds,
        disposeImmediately: Bool = false // ✅ NEW: Aggressive disposal flag
    ) async throws -> Bool {
        let startTime = Date()
        let wasEngineAlive = isEngineAlive

        // Ensure engine is initialized
        try await ensureEngineInitialized()

        let initTime = Date().timeIntervalSince(startTime)
        if !wasEngineAlive {
            print("FlutterEngineManager: Cold start - Engine initialized in \(Int(initTime * 1000))ms")
        } else {
            print("FlutterEngineManager: Warm start - Engine already alive")
        }

        // Execute callback with timeout
        return try await withThrowingTaskGroup(of: Bool.self) { group in
            // Task 1: Execute callback
            group.addTask {
                try await self.invokeCallback(callbackHandle: callbackHandle, input: input)
            }

            // Task 2: Timeout
            group.addTask {
                try await Task.sleep(nanoseconds: UInt64(timeoutSeconds * 1_000_000_000))
                throw FlutterEngineError.timeout
            }

            // Return first result (either success or timeout)
            guard let result = try await group.next() else {
                throw FlutterEngineError.unknown
            }

            group.cancelAll()

            let totalTime = Date().timeIntervalSince(startTime)
            print("FlutterEngineManager: Callback (handle: \(callbackHandle)) completed in \(Int(totalTime * 1000))ms")

            // ✅ NEW: Aggressive disposal logic
            if disposeImmediately {
                print("🔥 Aggressive disposal: Killing engine immediately to free RAM")
                dispose()
            } else {
                // Original behavior: Keep engine alive for 5 minutes
                queue.sync {
                    self.lastUsedTimestamp = Date()
                }
                scheduleDisposalCheck()
            }

            return result
        }
    }

    /// Dispose the Flutter Engine and free resources.
    ///
    /// WARNING: This will make next callback execution slow (cold start).
    /// Only call this if memory is critical.
    func dispose() {
        queue.sync {
            print("FlutterEngineManager: Disposing engine...")

            // ✅ FIX: Remove method channel handler BEFORE clearing reference
            // This prevents handler retention and ensures proper cleanup
            methodChannel?.setMethodCallHandler(nil)
            methodChannel = nil

            engine = nil
            isInitialized = false

            // ✅ NEW: Cancel any pending disposal checks
            disposalWorkItem?.cancel()
            disposalWorkItem = nil
            lastUsedTimestamp = nil

            print("FlutterEngineManager: Engine disposed")
        }
    }

    // MARK: - Private Methods

    /// Ensure Flutter Engine is initialized (thread-safe, async).
    private func ensureEngineInitialized() async throws {
        // Fast path: already initialized
        if isInitialized {
            return
        }

        // Slow path: need initialization
        return try await withCheckedThrowingContinuation { continuation in
            queue.async {
                // Double-check after acquiring lock
                if self.isInitialized {
                    continuation.resume()
                    return
                }

                // If already initializing, queue this continuation
                if self.isInitializing {
                    self.initializationContinuations.append(continuation)
                    return
                }

                // Start initialization
                self.isInitializing = true
                self.initializationContinuations.append(continuation)

                // Perform initialization on main thread (required by Flutter)
                DispatchQueue.main.async {
                    self.performInitialization()
                }
            }
        }
    }

    /// Perform actual engine initialization (must be called on main thread).
    private func performInitialization() {
        guard let callbackHandle = self.callbackHandle else {
            let error = FlutterEngineError.noCallbackHandle
            completeInitialization(error: error)
            return
        }

        print("FlutterEngineManager: Initializing Flutter Engine...")

        // Create engine
        let engine = FlutterEngine(name: "native_workmanager_background")

        // Run with callback dispatcher entry point
        let callbackInfo = FlutterCallbackCache.lookupCallbackInformation(callbackHandle)
        guard let callbackInfo = callbackInfo else {
            let error = FlutterEngineError.invalidCallbackHandle
            completeInitialization(error: error)
            return
        }

        let success = engine.run(withEntrypoint: callbackInfo.callbackName,
                                libraryURI: callbackInfo.callbackLibraryPath)

        if !success {
            let error = FlutterEngineError.engineStartFailed
            completeInitialization(error: error)
            return
        }

        // Setup method channel
        let channel = FlutterMethodChannel(
            name: "dev.brewkits/dart_worker_channel",
            binaryMessenger: engine.binaryMessenger
        )

        self.engine = engine
        self.methodChannel = channel

        // Wait for Dart to signal ready
        waitForDartReady(channel: channel, timeout: FlutterEngineManager.initTimeoutSeconds)
    }

    /// Wait for Dart side to signal it's ready.
    private func waitForDartReady(channel: FlutterMethodChannel, timeout: TimeInterval) {
        var isReady = false
        let startTime = Date()

        channel.setMethodCallHandler { [weak self] (call, result) in
            guard let self = self else { return }

            if call.method == "dartReady" {
                isReady = true
                result(nil)

                print("FlutterEngineManager: Dart side ready")

                self.queue.async {
                    self.isInitialized = true
                    self.isInitializing = false

                    // Resume all waiting continuations
                    let continuations = self.initializationContinuations
                    self.initializationContinuations.removeAll()

                    for continuation in continuations {
                        continuation.resume()
                    }
                }
            }
        }

        // Timeout check
        DispatchQueue.global().asyncAfter(deadline: .now() + timeout) { [weak self] in
            guard let self = self else { return }

            if !isReady {
                print("FlutterEngineManager: Timeout waiting for Dart ready signal")
                self.completeInitialization(error: FlutterEngineError.dartReadyTimeout)
            }
        }
    }

    /// Complete initialization with error.
    private func completeInitialization(error: Error) {
        queue.async {
            self.isInitializing = false

            let continuations = self.initializationContinuations
            self.initializationContinuations.removeAll()

            for continuation in continuations {
                continuation.resume(throwing: error)
            }
        }
    }

    /// Invoke Dart callback via method channel.
    private func invokeCallback(callbackHandle: Int64, input: String?) async throws -> Bool {
        guard let channel = methodChannel else {
            throw FlutterEngineError.engineNotInitialized
        }

        return try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.main.async {
                // ✅ FIXED: Pass callbackHandle (not callbackId) to background isolate
                // The background isolate will resolve the handle to the actual callback function
                let args: [String: Any?] = [
                    "callbackHandle": callbackHandle,
                    "input": input
                ]

                channel.invokeMethod("executeCallback", arguments: args) { result in
                    if let error = result as? FlutterError {
                        print("FlutterEngineManager: Callback error: \(error.message ?? "unknown")")
                        continuation.resume(returning: false)
                    } else if let success = result as? Bool {
                        continuation.resume(returning: success)
                    } else {
                        // Default to true if result is nil (backward compatibility)
                        continuation.resume(returning: true)
                    }
                }
            }
        }
    }

    // ✅ IMPLEMENTED: Auto-disposal implementation
    /// Schedule automatic engine disposal after idle timeout.
    ///
    /// This method schedules a check to dispose the engine if it has been idle
    /// for longer than `idleTimeoutSeconds`. Each new callback execution resets
    /// the timer.
    private func scheduleDisposalCheck() {
        queue.async { [weak self] in
            guard let self = self else { return }

            // Cancel any existing disposal check
            self.disposalWorkItem?.cancel()

            // Create new disposal work item
            let workItem = DispatchWorkItem { [weak self] in
                guard let self = self else { return }

                self.queue.sync {
                    // Check if engine is still idle
                    guard let lastUsed = self.lastUsedTimestamp else {
                        return
                    }

                    let idleTime = Date().timeIntervalSince(lastUsed)
                    if idleTime >= FlutterEngineManager.idleTimeoutSeconds {
                        print("FlutterEngineManager: Auto-disposing after \(Int(idleTime))s idle")
                        self.dispose()
                    }
                }
            }

            self.disposalWorkItem = workItem

            // Schedule disposal check
            DispatchQueue.global(qos: .utility).asyncAfter(
                deadline: .now() + FlutterEngineManager.idleTimeoutSeconds,
                execute: workItem
            )
        }
    }
}

// MARK: - Errors

enum FlutterEngineError: LocalizedError {
    case noCallbackHandle
    case invalidCallbackHandle
    case engineStartFailed
    case dartReadyTimeout
    case engineNotInitialized
    case timeout
    case unknown

    var errorDescription: String? {
        switch self {
        case .noCallbackHandle:
            return "No callback handle set. Call setCallbackHandle() during initialization."
        case .invalidCallbackHandle:
            return "Invalid callback handle. Make sure dispatcher is a top-level function."
        case .engineStartFailed:
            return "Failed to start Flutter Engine."
        case .dartReadyTimeout:
            return "Timeout waiting for Dart ready signal."
        case .engineNotInitialized:
            return "Engine not initialized."
        case .timeout:
            return "Callback execution timeout."
        case .unknown:
            return "Unknown error."
        }
    }
}
